Add alternative GCP workload identity impersonation setup (#3088)
astronomer/docs: astro/authorize-deployments-to-your-cloud.md

title: 'Authorize an Astro Deployment to cloud resources using workload identity'
id: authorize-deployments-to-your-cloud
description: Give Astro Deployments access to your cloud resources using a Kubernetes workload identity

> toc_max_heading_level: 3

Author: ```suggestion
toc_max_heading_level: 2
```

    ]}>
<TabItem value="aws">

> ### Step 1: Authorize the Deployment in your cloud

Reviewer: ```suggestion
#### Step 1: Authorize the Deployment in your cloud
```

Repeat these steps for each Astro Deployment that needs to access your AWS resources.

> ### Step 2: Create an Airflow connection

Reviewer: ```suggestion
#### Step 2: Create an Airflow connection
```

<TabItem value="gcp">

> #### Authorize your Deployment through GCP Service Account Impersonation

Author: Yeah it's a tab thing, but there's a better solution here and I'm being a bit lazy!

[GCP service account impersonation](https://cloud.google.com/docs/authentication/use-service-account-impersonation) allows your Deployment workload identity to assume an existing service account on your GCP project. This is the most secure authorization setup because your Deployment only uses generated, short-lived credentials for a service account, rather than a persistent and static service account key. 

1. [Create a service account](https://cloud.google.com/iam/docs/service-accounts-create) in the GCP project that you want your Deployment to access. Grant the service account any permissions that the Deployment will need in your GCP project. Copy the service account ID to use later in this setup.
2. In the Cloud UI, select your Deployment, then click **Details**. Copy the Deployment's **Workload Identity**.
3. In the Google Cloud Console, open the **IAM & admin** menu, then open the service account you just created. 

> 4. In the **Actions** column, click **Manage Permissions**, then click **Grant Access**. In the window that appears, enter your Deployment's workload identity ID in the **Add Principals** field and select the [`Service Account Token Creator`](https://cloud.google.com/iam/docs/understanding-roles#iam.serviceAccountTokenCreator) in the **Assign Roles** field.

Reviewer 2: ```suggestion
4. In the **Actions** column, click **Manage Permissions**, then click **Grant Access**. In the window that appears, enter your Deployment's workload identity service account in the **Add Principals** field and select the [`Service Account Token Creator`](https://cloud.google.com/iam/docs/understanding-roles#iam.serviceAccountTokenCreator) in the **Assign Roles** field.
```

> 4. In the **Actions** column, click **Manage Permissions**, then click **Grant Access**. In the window that appears, enter your Deployment's workload identity service account in the **Add Principals** field and select the [`Service Account Token Creator`](https://cloud.google.com/iam/docs/understanding-roles#iam.serviceAccountTokenCreator) in the **Assign Roles** field.

Author: It's not a full new window, so I like using modal here

        
    Note that this implementation requires `apache-airflow-providers-google >= 10.8.0`. See [Add Python, OS-level packages, and Airflow providers](https://docs.astronomer.io/astro/cli/develop-project#add-python-os-level-packages-and-airflow-providers).

   - Specify the impersonation chain in code when you instantiate a Google Cloud operator. See [Airflow documentation](https://airflow.apache.org/docs/apache-airflow-providers-google/stable/connections/gcp.html#direct-impersonation-of-a-service-account). Note that if you configure both a connection type and an operator, the operator-level configuration takes precedence.
    - To access resources in a secrets backend, run the following command to create an environment variable that grants access to the secrets backend:

    ```zsh
    astro deployment variable create --deployment-id <your-deployment-id> AIRFLOW__SECRETS__BACKEND_KWARGS={"connections_prefix": "airflow-connections", "variables_prefix": "airflow-variables", "project_id": "<your-secret-manager-project-id>", "impersonation_chain": "<your-gcp-service-account>"}
    ```

#### Alternative setup: Grant an IAM role to your Deployment workload identity

> Complete this alternative, simpler setup if you can tolerate your Deployment having long-lasting credentials to your Google Cloud project.

Reviewer 2: also, i wouldn't say this has the downside of long-lasting credentials; technically that's the case with both. maybe just say something like "use this option if you don't have an existing Google service account you'd like your Astro Workload Identity to impersonate"

> Complete this alternative, simpler setup if you can tolerate your Deployment having long-lasting credentials to your Google Cloud project.

Author: ```suggestion
Complete this alternative setup if you don't have an existing Google service account that your Deployment workload identity can impersonate.
```

> Complete this alternative setup if you want a more simple setup and can tolerate your Deployment having long-lasting credentials to your Google Cloud project.

Author: ```suggestion
Complete this alternative, simpler setup if you can tolerate your Deployment having long-lasting credentials to your Google Cloud project.
```

<TabItem value="gcp">

> ### Authorize your Deployment through GCP Service Account Impersonation

Reviewer: Ok so I think we do have to have these headers as H4 - right now, half of the right nav just doesn't really work depending on which tab you're on (AWS or GCP). What a strange docusaurus quirk. I think they either have to be H3 and identical between tabs, or H4 -
<img width="289" alt="Screenshot 2023-11-17 at 8 09 25 AM" src="https://github.com/astronomer/docs/assets/25642903/7cc886ef-66e3-42da-ab4e-0444a19e0956">

Author: Oh I think I need to lower the frontmatter requirement by 1 level, then it should work

> ### Authorize your Deployment through GCP Service Account Impersonation

Reviewer: ```suggestion
#### Authorize your Deployment through GCP Service Account Impersonation
```

> [GCP service account impersonation](https://cloud.google.com/docs/authentication/use-service-account-impersonation) allows your Deployment workload identity to assume an existing service account on your GCP project. This is the most secure authorization setup because your Deployment only uses generated, short-lived credentials for a service account, rather than a persistent and static service account key. 

Reviewer 3: ```suggestion
[GCP service account impersonation](https://cloud.google.com/docs/authentication/use-service-account-impersonation) allows your Deployment's workload identity to assume an existing service account in your GCP project. This is the most secure authorization setup because your Deployment only uses generated, short-lived credentials for a service account, rather than a persistent and static service account key. 
```

1. [Create a service account](https://cloud.google.com/iam/docs/service-accounts-create) in the GCP project that you want your Deployment to access. Grant the service account any permissions that the Deployment will need in your GCP project. Copy the service account ID to use later in this setup.
2. In the Cloud UI, select your Deployment, then click **Details**. Copy the Deployment's **Workload Identity**.

> 3. In the Google Cloud Console, open the **IAM & admin** menu, then open the service account you just created. 

Reviewer 3: ```suggestion
3. In the Google Cloud Console, open the **IAM & Admin > Service Accounts** menu, then open the service account you just created. 
```

        
    Note that this implementation requires `apache-airflow-providers-google >= 10.8.0`. See [Add Python, OS-level packages, and Airflow providers](https://docs.astronomer.io/astro/cli/develop-project#add-python-os-level-packages-and-airflow-providers).

   - Specify the impersonation chain in code when you instantiate a Google Cloud operator. See [Airflow documentation](https://airflow.apache.org/docs/apache-airflow-providers-google/stable/connections/gcp.html#direct-impersonation-of-a-service-account). Note that if you configure both a connection type and an operator, the operator-level configuration takes precedence.
    - To access resources in a secrets backend, run the following command to create an environment variable that grants access to the secrets backend:

    ```zsh
    astro deployment variable create --deployment-id <your-deployment-id> AIRFLOW__SECRETS__BACKEND_KWARGS={"connections_prefix": "airflow-connections", "variables_prefix": "airflow-variables", "project_id": "<your-secret-manager-project-id>", "impersonation_chain": "<your-gcp-service-account>"}
    ```

> ### Alternative setup: Grant an IAM role to your Deployment workload identity

Reviewer: ```suggestion
#### Alternative setup: Grant an IAM role to your Deployment workload identity
```

3. Click **Save**. 
    
    If you don't see **Google Cloud** as a connection type, ensure you have installed its provider package in your Astro project's `requirements.txt` file. See **Use Provider** in the [Astronomer Registry](https://registry.astronomer.io/providers/Google/versions/latest) for the latest package.

#### Alternative setup: Use GCP Service Account Impersonation

> [GCP service account impersonation](https://cloud.google.com/docs/authentication/use-service-account-impersonation) allows your Deployment workload identity to assume an existing service account on your GCP project. Complete this setup if you want your Deployment to only use short-lived credentials to connect to your cloud, rather than a dedicated IAM role. 

Author: @chomipi88 Is the main reason you would do this alternative setup? If not, what's the primary benefit here?

Reviewer 3: made a suggestion below

> [GCP service account impersonation](https://cloud.google.com/docs/authentication/use-service-account-impersonation) allows your Deployment workload identity to assume an existing service account on your GCP project. Complete this setup if you want your Deployment to only use short-lived credentials to connect to your cloud, rather than a dedicated IAM role. 

Reviewer 3: ```suggestion
[GCP service account impersonation](https://cloud.google.com/docs/authentication/use-service-account-impersonation) allows your Deployment workload identity to assume an existing service account on your GCP project. Complete this setup if you want your Deployment to only use its authenticated identity to generate short-lived credentials for a service account that has the authorization that your use case requires, rather than persistent and static service account key. 
```

> 1. [Create a service account](https://cloud.google.com/iam/docs/service-accounts-create) in the Google project that you want your Deployment to access. Copy the service account ID to user later in this setup.

Reviewer 3: Also requires granting this service account the necessary permissions to the target service (eg. GCS Bucket)

> 1. [Create a service account](https://cloud.google.com/iam/docs/service-accounts-create) in the Google project that you want your Deployment to access. Copy the service account ID to user later in this setup.

Author: ```suggestion
1. [Create a service account](https://cloud.google.com/iam/docs/service-accounts-create) in the GCP project that you want your Deployment to access. Grant the service account any permissions that the Deployment will need in your GCP project. Copy the service account ID to user later in this setup.
```

2. In the Cloud UI, select your Deployment, then click **Details**. Copy the Deployment's **Workload Identity**.

> 3. Grant the [`Service Account Token Creator`](https://cloud.google.com/iam/docs/understanding-roles#iam.serviceAccountTokenCreator) IAM role to the principal account, also known as your Deployment's workload identity. See [Grant a single role.](https://cloud.google.com/iam/docs/granting-changing-revoking-access#grant-single-role).

Reviewer 3: Specifically the user needs to grant this role to the Deployment's workload identity TO the service account they created in step 1 (or an existing SA).

Author: How would you limit it so that it can only generate a token for the specific SA? Is there a specific [condition](https://cloud.google.com/iam/docs/conditions-overview) we should tell them to add?

Reviewer 3: The specific condition is granting the Astro Deployment WI the `Service Account Token Creator` role on their specific target SA.
This means the Astro Deployment WI can **only** generate short lived tokens for the specific target SA.

Author: Right, I was trying to figure out how to actually apply that condition. [The GCP docs](https://cloud.google.com/docs/authentication/use-service-account-impersonation) only mention the role, but not the limiting condition

Reviewer 3: Ah its a matter of granting the principal (Astro Deployment WI) the role mentioned to the service account.
https://cloud.google.com/iam/docs/manage-access-service-accounts?_ga=2.156967740.-620980664.1671590357
<img width="1574" alt="image" src="https://github.com/astronomer/docs/assets/87984472/d852f77d-9ad2-4fa7-8831-c76324e8426d">

Author: This screenshot was a life saver, just updated the setup to lead users exactly here.

> 3. Grant the [`Service Account Token Creator`](https://cloud.google.com/iam/docs/understanding-roles#iam.serviceAccountTokenCreator) IAM role to the principal account, also known as your Deployment's workload identity. See [Grant a single role.](https://cloud.google.com/iam/docs/granting-changing-revoking-access#grant-single-role).

Reviewer 3: ```suggestion
3. Grant the [`Service Account Token Creator`](https://cloud.google.com/iam/docs/understanding-roles#iam.serviceAccountTokenCreator) IAM role to the principal account, also known as your Deployment's workload identity. See [Grant a single role](https://cloud.google.com/iam/docs/granting-changing-revoking-access#grant-single-role).
```

4. Complete one of the following options for your Deployment to access your cloud resources:

    - Create a **Google Cloud** connection type in Airflow and configure the following values:
        - **Connection Id**: Enter a name for the connection.

>         - **Impersonation Chain**: Enter the ID of the service account that your Deployment should impersonate.

Author: ```suggestion
        - **Impersonation Chain**: Enter the ID of the service account that your Deployment should impersonate.
   - Specify the impersonation chain at the operator level. See [Airflow documentation](https://airflow.apache.org/docs/apache-airflow-providers-google/stable/connections/gcp.html#direct-impersonation-of-a-service-account). Note that if you configure both a connection type and an operator, the operator-level configuration takes precedence.
```

>     - Run the following command to create an environment variable that connects your Deployment to your GCP secrets backend:

Author: @chomipi88 I didn't know how to add the "Leverage the Astro Workload Identity" option here, lmk the specific steps I should add ([based on the content in the original issue](https://github.com/astronomer/docs/issues/2922))

Reviewer 3: The first option is configuring the impersonation chain on the connection as per: https://airflow.apache.org/docs/apache-airflow-providers-google/stable/connections/gcp.html#configuring-the-connection:~:text=request%20only%20once.-,Impersonation%20Chain,-Optional%20service%20account

The second option is configuring the impersonation chain on the operator as per: https://airflow.apache.org/docs/apache-airflow-providers-google/stable/connections/gcp.html#direct-impersonation-of-a-service-account
Important to note that the `impersonation_chain` passed to the operator takes precedence.

>     - Run the following command to create an environment variable that connects your Deployment to your GCP secrets backend:

Author: ```suggestion
    - To access resources in a secrets backend, run the following command to create an environment variable that grants access to the secrets backend:
```

    ```zsh

>     astro deployment variable create --deployment-id <your-deployment-id> AIRFLOW__SECRETS__BACKEND_KWARGS={"connections_prefix": "airflow-connections", "variables_prefix": "airflow-variables", "project_id": "<your-secret-manager-project-id>", "impersonation_chain": "<your-gcp-service-account>"}

Author: Does this command only grant you access to the secrets backend, or is this a general setup to access all resources in the GCP project?

Reviewer 3: This third option is to configure the secrets backend for a particular deployment where connections and variables could be stored.

Reviewer 2: we already have [separate instructions for Google secrets backend](https://deploy-preview-3088--gallant-neumann-56599d.netlify.app/astro/secrets-backend/gcp-secretsmanager#step-3-configure-secret-manager-on-astro-using-workload-identity-recommended), do we want to duplicate that here?

3. Click **Save**. 
    
    If you don't see **Google Cloud** as a connection type, ensure you have installed its provider package in your Astro project's `requirements.txt` file. See **Use Provider** in the [Astronomer Registry](https://registry.astronomer.io/providers/Google/versions/latest) for the latest package.

> #### Alternative setup: Use GCP Service Account Impersonation

Reviewer 2: maybe it doesn't matter, but do we want to recommend this as an alternative or the recommended way?

Reviewer 3: Given the coming work on pre-defined, I do agree that we should start positioning impersonation as the recommended way.

Author: Based on the language around it I assumed it was the alternative, but happy to frame it as the default method

Reviewer 3: @jwitz yes pls let's do that. should of called that out sooner... my bad.

[GCP service account impersonation](https://cloud.google.com/docs/authentication/use-service-account-impersonation) allows your Deployment workload identity to assume an existing service account on your GCP project. Complete this setup if you want your Deployment to only use generated, short-lived credentials for a service account, rather than a persistent and static service account key. 

> 1. [Create a service account](https://cloud.google.com/iam/docs/service-accounts-create) in the GCP project that you want your Deployment to access. Grant the service account any permissions that the Deployment will need in your GCP project. Copy the service account ID to user later in this setup.

Reviewer 2: typo: `to user later`

2. In the Cloud UI, select your Deployment, then click **Details**. Copy the Deployment's **Workload Identity**.

> 3. Grant the [`Service Account Token Creator`](https://cloud.google.com/iam/docs/understanding-roles#iam.serviceAccountTokenCreator) IAM role to the principal account, also known as your Deployment's workload identity, so that your Deployment can impersonate your service account. See [Grant a single role](https://cloud.google.com/iam/docs/granting-changing-revoking-access#grant-single-role).

Reviewer 2: what do we mean by "principal account"? that's the first time using this term; maybe we just say "to the service account you just created"?

it might be better to just link to the impersonation docs? https://cloud.google.com/docs/authentication/use-service-account-impersonation#required-roles

Author: I thought the "principal" is the _impersonating_ service account, which in this case would be the Deployment workload identity. I used "principal" here because it's in the GCP docs and because it distinguishes the Deployment service account from the service account being impersonated.

Author: Maybe it makes more sense to start with the Deployment language first and then clarify with the principal language, see my suggestion below

Reviewer 3: Jake is right. The principal would be the Deployment WI.

Reviewer 2: i mean that's my point, i'm not sure that terminology is well-understood

> 3. Grant the [`Service Account Token Creator`](https://cloud.google.com/iam/docs/understanding-roles#iam.serviceAccountTokenCreator) IAM role to the principal account, also known as your Deployment's workload identity, so that your Deployment can impersonate your service account. See [Grant a single role](https://cloud.google.com/iam/docs/granting-changing-revoking-access#grant-single-role).

Author: ```suggestion
3. Grant the [`Service Account Token Creator`](https://cloud.google.com/iam/docs/understanding-roles#iam.serviceAccountTokenCreator) IAM role to your Deployment's workload identity, also known as the _principal_ service account. See [Grant a single role](https://cloud.google.com/iam/docs/granting-changing-revoking-access#grant-single-role).
```

> 2. In the Cloud UI, select your Deployment, then click **Details**. Copy the Deployment's **Workload Identity**.
> 3. Grant the [`Service Account Token Creator`](https://cloud.google.com/iam/docs/understanding-roles#iam.serviceAccountTokenCreator) IAM role to the principal account, also known as your Deployment's workload identity, so that your Deployment can impersonate your service account. See [Grant a single role](https://cloud.google.com/iam/docs/granting-changing-revoking-access#grant-single-role).

Author: ```suggestion
2. In the Cloud UI, select your Deployment, then click **Details**. Copy the Deployment's **Workload Identity**.
3. In the Google Cloud Console, open the **IAM & admin** menu, then open the service account you just created. 
4. In the **Actions** column, click **Manage Permissions**, then click **Grant Access**. In the window that appears, enter your Deployment's workload identity ID in the **Add Principals** field and select the [`Service Account Token Creator`](https://cloud.google.com/iam/docs/understanding-roles#iam.serviceAccountTokenCreator) in the **Assign Roles** field.
```

4. Complete one of the following options for your Deployment to access your cloud resources:

    - Create a **Google Cloud** connection type in Airflow and configure the following values:
        - **Connection Id**: Enter a name for the connection.

>         - **Impersonation Chain**: Enter the ID of the service account that your Deployment should impersonate.

Reviewer 2: might be worth mentioning they need Google provider >= 10.8.0 for this to work

>         - **Impersonation Chain**: Enter the ID of the service account that your Deployment should impersonate.

Author: ```suggestion
        - **Impersonation Chain**: Enter the ID of the service account that your Deployment should impersonate.
        
        Note that this implementation requires `apache-airflow-providers-google >= 10.8.0`. See [Add Python, OS-level packages, and Airflow providers](https://docs.astronomer.io/astro/cli/develop-project#add-python-os-level-packages-and-airflow-providers).
```

>    - Specify the impersonation chain at the operator level. See [Airflow documentation](https://airflow.apache.org/docs/apache-airflow-providers-google/stable/connections/gcp.html#direct-impersonation-of-a-service-account). Note that if you configure both a connection type and an operator, the operator-level configuration takes precedence.

Reviewer 2: instead of saying "at the operator level", probably better to say "in code when instantiating a Google Cloud Operator"

>    - Specify the impersonation chain at the operator level. See [Airflow documentation](https://airflow.apache.org/docs/apache-airflow-providers-google/stable/connections/gcp.html#direct-impersonation-of-a-service-account). Note that if you configure both a connection type and an operator, the operator-level configuration takes precedence.

Author: ```suggestion
   - Specify the impersonation chain in code when you instantiate a Google Cloud operator. See [Airflow documentation](https://airflow.apache.org/docs/apache-airflow-providers-google/stable/connections/gcp.html#direct-impersonation-of-a-service-account). Note that if you configure both a connection type and an operator, the operator-level configuration takes precedence.
```
